ReactAndroid/src/main/java/com/facebook/react/modules/core/Timing.java

```diff
@@ -107,13 +112,78 @@ public void doFrame(long frameTimeNanos) {
     }
   }
 
+  private class IdleFrameCallback implements Choreographer.FrameCallback {
+
+    @Override
+    public void doFrame(long frameTimeNanos) {
+      if (isPaused.get()) {
+        return;
+      }
+
+      // If the JS thread is busy for multiple frames we cancel any other pending runnable.
+      if (mCurrentIdleCallbackRunnable != null) {
+        mCurrentIdleCallbackRunnable.cancel();
+      }
+
+      long frameTimeMillis = frameTimeNanos / 1000000;
+      long timeSinceBoot = SystemClock.uptimeMillis();
+      long frameTimeElapsed = timeSinceBoot - frameTimeMillis;
+      long time = SystemClock.currentTimeMillis();
+      long absoluteFrameStartTime = time - frameTimeElapsed;
+
+      mCurrentIdleCallbackRunnable = new IdleCallbackRunnable(absoluteFrameStartTime);
+      getReactApplicationContext().runOnJSQueueThread(mCurrentIdleCallbackRunnable);
+
+      Assertions.assertNotNull(mReactChoreographer).postFrameCallback(
+          ReactChoreographer.CallbackType.IDLE_EVENT,
+          this);
+    }
+  }
+
+  private class IdleCallbackRunnable implements Runnable {
+    private volatile boolean mCanceled = false;
+    private final long mFrameStartTime;
+
+    public IdleCallbackRunnable(long frameStartTime) {
+      mFrameStartTime = frameStartTime;
+    }
+
+    @Override
+    public void run() {
+      if (mCanceled) {
+        return;
+      }
+
+      List<ExecutorToken> contextsToCall;
+      synchronized (mIdleCallbackGuard) {
+        contextsToCall = new ArrayList<>(mSendIdleEventsExecutorTokens);
```

Comment:
let's try to avoid this allocation on every frame. We can just keep one array and copy into it